java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/TextDocumentServiceImpl.java

```diff
@@ -1088,6 +1121,39 @@ public Void visitMethod(MethodTree tree, Void p) {
         return result;
     }
 
+    private List<Pair<String, String>> getProjectConfigurations(JavaSource source) {
+        for (FileObject fo : source.getFileObjects()) {
+            Project p = FileOwnerQuery.getOwner(fo);
+            if (p != null) {
+                ProjectConfigurationProvider<ProjectConfiguration> provider = p.getLookup().lookup(ProjectConfigurationProvider.class);
+                List<Pair<String, String>> configDispNames = new ArrayList<>();
+                if (provider != null) {
+                    boolean skippedFirst = false;
+                    for (ProjectConfiguration c : provider.getConfigurations()) {
+                        if (skippedFirst) {
+                            String runConfig = null;
+                            String debugConfig = null;
+                            if (provider.configurationAffectsAction(c, ActionProvider.COMMAND_RUN_SINGLE)) {
```

Comment:
I am not sure about this approach. If a Configuration does not define an action mapping, Maven and Gradle will fall back to the default (implicit) configuration's behaviour. This condition seems to disable the action if a configuration does not define a specialization for it. I think the semantic for LSP client diverges from the semantic seen by the IDE / headless LSP server.

Reply:
OK, one more thought. The code here seems to check if an action (run, debug) is **available** in a specific configuration. We have a similar issue with plain "Run" action in vscode (which means essentially "Debug" in vscode terminology). It was planned to display a warning message, something like "Debug is not supported in XXX config, running without debugger" - and for that I need a similar info.

What if, instead of `configurationAffectsAction` (or in addition to), `isActionEnabled(Configuration, String)` is introduced ? We can **enhance** Maven and Gradle XML format with `disable` flag for the actions (or autodetect when e.g. commandline or goal list is empty ?), and finally declare `debug` as disabled in the `mn:run` or `gradle --continuous` case. That would work for both code lens, the regular debug adapter, and would be consistent with the IDE behaviour.

@JaroslavTulach, @MartinBalin, @entlicher -- opinions ?

Reply:
There's no point in falling back. I intentionally do not want that.
The intention is to select unique Run/Debug actions from configurations. And this intention conveniently filters Debug in Continuous mode out. There's no point adding duplicit actions that would fall back somewhere into Code Lenses. Choosing simple Run/Debug is the fall-back already.

Reply:
My point was, that the project support(s) DO such fallback, and the API is defined so that configuration is just an override (the concept of disabling an action in a configuration is not present in the implementation, that's right). 

So the point is that this approach works now, in the current de-facto state, for these specific actions. But can break in the future - consider (some) Configuration that focuses on other technology aspect than `--continuous`, i.e. on build or package; if it won't defne `run.single`, Run would be disabled.

IMHO it would be better to somehow implement @JaroslavTulach's suggestion to allow plugins (maven, gradle) to indicate the action is to be disabled in a configuration - and propagate that from API through `isActionEnabled` query.

Reply:
O.K. it makes sense. It's better not to have the fallback to the default configuration in the action impl, but just in the projects infrastructure, I agree.

Reply:
OK, if you agree, I'll prepare a PR that allows to disable an action that for maven + gradle. 

Do you think we should disable `debug` for the 'continuous' mode in the IDE as well ? Right now running Debug in dev mode will just run the debugger, without 'continuous' parameter.

Reply:
Yes, thanks, I agree.
In the IDE I'm not sure. As the user does Debug explicitly, it's probably better to start debugger, but it'd be good to show a warning that we do not use the Continuous Mode configuration, I think.
In VSCode I didn't find a way how to switch to plain run when debug is started. We probably need to cancel debugger and auto-start Run.